### PR TITLE
git info in state name

### PR DESCRIPTION
**Pull request by stevekrouse**
This allows us to have separate "database forks" for each branch, which I think is the behavior we want for now.

Once we need to do an atomic migration https://github.com/compose-run/server/issues/31, we'll need to update this scheme a bit to handle merging to main. Now sure how we'll do that at the moment, but I think this is ok to merge for now.

@parenthetical - thoughts?

### CODE REVIEW

File: src/state/appName.ts [hunk 1]
```diff
@@ -1,2 +1,2 @@
-const appName = "@compose-run/community";
+const appName = `@${process.env.REACT_APP_VERCEL_GIT_REPO_OWNER}/${process.env.REACT_APP_VERCEL_GIT_REPO_SLUG}/${process.env.REACT_APP_VERCEL_GIT_COMMIT_REF}`;
```
Comment (parenthetical) [inline src/state/appName.ts:1]: Why does the commit ref go in here? Wouldn't that reset the state every time?

**Comment by stevekrouse** [comment]
That's the branch name, not the commit hash. It's confusingly named.

File: env.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  echo REACT_APP_VERCEL_GIT_REPO_OWNER=compose-run
+  echo REACT_APP_VERCEL_GIT_REPO_SLUG=community 
+  echo REACT_APP_VERCEL_GIT_COMMIT_REF=$(git symbolic-ref --short HEAD)
+  echo REACT_APP_VERCEL_GIT_COMMIT_SHA=$(git log --pretty=format:'%H' -n 1)
+} > .env.development.local
```
Comment (parenthetical) [inline env.sh:6]: Maybe `>>`? Or another solution to prevent accidental overwrites of other things in there?

**Comment by parenthetical** [comment]
> That's the branch name, not the commit hash. It's confusingly named.

Oh sorry! That makes sense then. Merging back into main with incompatible state would be the hard bit. Also, for testing out migrations you'd want the state of the fork to come with you to the new branch, right?

File: env.sh (hunk 2 above)
Comment (stevekrouse) [reply env.sh:6]: My thought is that we'll reserve `.env.development.local` for this one purpose in this repo, so there's no risk of overwritting anything. There are a lot of alternative `.env` file names (below) that we can use if we need more env variables for this project. (Or we can just put them in this `.sh` file if they're meant to be for local development.)

Screen Shot 2021-12-08 at 2 53 29 PM

**Comment by stevekrouse** [comment]
> for testing out migrations you'd want the state of the fork to come with you to the new branch, right?

I imagine that you would test our the migrations in your branch like this:

```
name: `${appName}/messages-1` // where you increment the number each time you want to test your migration function
initialState: getCloudState('@compose-run/community/main@3a6611ae44c8cdacb6231671960e70ecd21f8ff6`).then(myMigrationFunc)
```

So you can test out the migration on the data from the main commit hash. (Or from any other place.)

> Merging back into main with incompatible state would be the hard bit

1.... My current scheme doesn't put the commit hash in the name. Maybe it should... We want it in the name for the main branch. If we only want it in the name for the main branch, we can accomplish that too...

2.... The only issue is that right before you merge, you have to ensure that your initial state's commit hash is the one of that's currently deployed.

3.... Or here's an alternate scheme where we don't need to put the commit hash anywhere: we can create a wrapper around `createCloudReducer` (so no change to our API) called `createMigratableCloudReducer` that accepts a `migration` field. It ensures that the migration is run exactly one time (on first load) for every name, but only if the person who loads it is the creator of the reducer. The way this would work under the hood is that it would actually store in the state which migrations have already been run on this piece of state, in say an array hashes of the migration function codes, in the state itself: `{ migrationsHashes: [ ... ], value: { ... } }`, where `value` is the actual value. It would then accept a "special action" which would trigger the migration function to run.

There are two downsides:

a) The writer of the pull request would have to load the code after merging to main. This isn't a big deal because a good pull request author already does this after a merge to confirm things are working.


b) Until the author does this check, any other user will have a broken experience... The solution here is that you have to merge pull requests locally (not via github), so you can pull up the state with the proper `main` name locally before deploying it. (This would require `createMigratableCloudReducer ` to detect when a client is using a pre-migrated version of the code -- by augmenting all actions with the hash of the last ran migration -- and throwing some sort of error that the application programer can expose to their users to get them to refresh the page.)


Thoughts?

**Comment by stevekrouse** [comment]
I'd also be open to using bidirectional lenses https://www.inkandswitch.com/cambria/ for the migration problem... It's weird but it could help with a lot of these coordination problems.

**Comment by stevekrouse** [comment]
I'm going to merge this. We can reopen this conversation in https://github.com/compose-run/server/issues/31 when we find the need to do an atomic merge

**Pull request merged**